### ISSUE TITLE
docker container changes ip upon restart

### ISSUE THREAD

**Issue by kiorky** [docker/docker#2801]
I used to have a docker version (maybe 0.6.2->0.6.5) where ip of a specific container stayed the same upon restarts.

Now, with 0.6.6, the ip changes at each container restart, is it a feature or a regression ?

/cc @jpetazzo 

CC from my email on docker dev :
We have had some recent threads about docker networking support and today Jamshid Afshar started another related thread with the same problematic users already have.

From my humble point of experimentations:
Today:
* At each restart containers ip change.
* Docker applies rules on the fly, and cannot have a way to reconfigure network mapping rules (PAT) on demand.

Those are big stoppers for other tools integration like shorewall which for example:
* Does not support PAT rules with hostname/dns inside
* Does not support custom named chains unless you write some perl. Personally, Perl is not my cup of tea.
* My solution as of now is to write some scripts to save/reload the docker conf, see http://www.shorewall.net/shorewall_extension_scripts.htm. Idea is to save docker chains and restore them at different stages of the shorewall starting process.
* In other words, without care if shorewall (re)starts after docker, it kills the docker networking.

A reliable and predictable ip (re)configuration system is a musthave feature i think.
* In an ideal world, we can have a way to statically assign ourselves the ip addresses of the containers.
* And more generally, in any case, we must have a way to tell docker to (re)configure its networking mappings & rules without restarting the whole docker daemon & containers (reapply the docker rules & postrouting stuff). This would prevent to write unreliable & prone to errors cruft in other tools to save/restore docker stuff.

**Comment by Xoib** [comment]
@ibukanov Can you share your scripts ? I can use it to make a more automatic and reliable workaround for #7807

**Comment by ibukanov** [comment]
@Xoib 

See https://github.com/ibukanov/ahome/tree/55c1ba708a4b47c7ec4c1d4d3fad7e4ccb90dd9b/dev/browser-thip . That assumes a writable /etc/hosts that comes with Docker 1.2 IIRC. There `start` is a bash script that is a default command for the container. It assumes that /info/web/ip comes from a read-only host volume shared with another container. That container, before starting the main process, writes own IP using the following shell command:

```
ip -o -f inet addr show eth0 | sed 's/.*\<inet\s\+\([^/]*\).*/\1/' > /info/web/ip
```

The start script runs in parallel with the main container process the code that updates /etc/hosts. The main process here is a VNC server with a browser session that I use to check a test deployment of a container, but it can be anything. The update code uses `inotifywait` to wait for changes in /info/web/ip and patches /etc/hosts accordingly. A slight complication is that to avoid running as root the script first makes /etc/hosts writable for the user account and then restarts itself under that user account.

**Comment by ibukanov** [comment]
@Xoib 

If you need in a container A to refer to a container B by the name that survives restarts of B, then there is simpler way to handle this than the solution from my previous comment. When B starts write its IP address and the name to a file like `hosts_for_A`on a host volume using /etc/hosts format like in
```
127.0.0.1 localhost
discovered-ip name-for-b
```
Then, when you run A, add `-v /host_dir/hosts_for_A:/etc/hosts` to the docker run command. This overrides Docker's management of /etc/hosts in the container by a custom file and allows to refer to the container B via name-for-b that will be updated each time B restarts.

**Comment by cpuguy83** [comment]
Updating of linked ip addresses is fixed on master.
There was a bug where if you used an alias for the linked in container, the ip wouldn't update properly, but if you used the same name for the link as the container's real name, it was ok.

Please test! https://master.dockerproject.com for a pre-compiler binary to test with!

This fix should be in 1.5

**Comment by anydoby** [comment]
+1

**Comment by brunoborges** [comment]
Can someone confirm if the fix mentioned by @cpuguy83 is really on 1.5 ?

**Comment by LK4D4** [comment]
@brunoborges Fix mentioned by @cpuguy83 in 1.5

**Comment by Xoib** [comment]
@LK4D4 What @brunoborges mean is does someone actually tested this specific use case with 1.5 ? As @cpuguy83 mentioned it is fixed within 1.5 branch.
I didn't had the [opportunity - free time] to test it myself. But thanks @ibukanov !

**Comment by christianparpart** [comment]
:+1:

I am on docker 1.5.0 and a docker restart still changes the IP, that is really awkward, because everything non-docker related needs to be adjusted upon every little restart.

Any plans to change this behaviour in any way?

**Comment by brunoborges** [comment]
:+1: thanks @christianparpart for testing this

**Comment by oscerd** [comment]
Are there news about the IP changes after docker restart?

**Comment by imton** [comment]
+1 Confirmed in 1.6 stills changing ip after restart

**Comment by cpuguy83** [comment]
Yes, this issue will be closed if this situation ever changes.

**Comment by alexfd** [comment]
+1 docker static ips.

I'm updating IPs in haproxy with

```
docker ps|awk '/\./ {print $NF}' | while read line; do
        sed -i "s#$(grep "server $line" haproxy.cfg)#server $line $(docker inspect --format '{{ .NetworkSettings.IPAddress }}' $line):80 check#g" haproxy.cfg
 done
````

where docker container names and server definitions in haproxy have identical names, so $line would be host.example.com.

**Comment by thaJeztah** [comment]
Depending on the use case, this could be partly resolved if users are able to manually provide an IP-address; see https://github.com/docker/docker/issues/6743 and the issue now opened in Libnetwork; https://github.com/docker/libnetwork/issues/161

**Comment by ibukanov** [comment]
A workaround that I ended up using was to add a dummy container with a process that just sleeps for ever and then use  --net=container:dummy with the original container. Then one can restart the original any number of times without changing IP.

**Comment by joune** [comment]
+1 for static container IPs

**Comment by flaviotruzzi** [comment]
+1

**Comment by phemmer** [comment]
wow. gordon is getting sassy :-)

**Comment by cristiklein** [comment]
+1 for `--net=container:CONTAINER_NET`

**Comment by phemmer** [comment]
@cristiklein see https://docs.docker.com/engine/userguide/networking/dockernetworks/#user-defined-networks

**Comment by aboch** [comment]
https://github.com/docker/docker/pull/19001 provides a way for user to explicitly specify the IP address(es) of the container during `docker run` and `docker network connect` on user-defined networks, and preserve them across container restart.
It is addressing the original issue https://github.com/docker/docker/issues/6743 on top of the new networking model brought in by libnetwork.

I think that addresses what is being asked in this thread, and we should close this issue ?

**Comment by thaJeztah** [comment]
thanks @aboch, I'll close this issue; but feel free to comment after it's closed

**Comment by brunoborges** [comment]
I think this issue asks for a way to indicate that, when an existing container is restarted, it should provide an option to keep the previous IP address, regardless of it being created on user-defined, or dynamic network.

**Comment by aboch** [comment]
@brunoborges
I went through the issue's initial description and earlier comments. I read:

From @kiorky:
> In an ideal world, we can have a way to statically assign ourselves the ip addresses of the containers.

From @GeorgeDewar:
> I would very much like to see the ability to specify an IP address on container startup (e.g. docker run -d --ip 10.0.0.123), or any other means of specifying a static IP address 

From @dashohoxha:
> I use docker containers as virtual machines. From my point of view, it is useful to be able to assign a fixed IP to a docker container.

At least originally the issue was opened asking for a way to statically assign an IP address to a container. 
And I agree with @thaJeztah in https://github.com/docker/docker/issues/2801#issuecomment-118098608 that the statically configurable IP address would solve some if not most of the use-cases I read in the thread.

Therefore I stand on closing this one now.

Regarding your requirement of preserving the IP address. If an app running in the container need to be aware of the container's IP address, it needs to work at Layer 3, then doesn't this mean this container is a good candidate to be assigned a static IP address ?

At the moment I am not convinced we need an extra `docker run` or `docker network connect` option to preserve the IP address across restart for two reasons:
- primarily because the container restart could fail like you said if the IP is no longer available and this would not be recoverable
- and because now we can configure it statically via `--ip`, `--ip6`

If you still think the preserve ip option makes sense, please open a dedicated issue in libnetwork for it, as it is not the same request this issue originally tracked.

Thanks

**Comment by alexfd** [comment]
@aboch your comment in [19001] (https://github.com/docker/docker/pull/19001) says:
> This PR will allow user to choose the IP address(es) for the container during docker run and docker network connect. **The configuration is persisted across container restart**.

So that mostly solves what @brunoborges said anyway.
Adding another flag would just be redundant.

**Comment by tommasop** [comment]
tommasop@thinkingco.de sent you an invitation to join MEGA.

--- Message from user: ---
Hello, join me on MEGA and get access to encrypted storage and
communication. Get 50 GB free!
---

Please create your free MEGA account by clicking on the following
link:
https://mega.nz/#newsignupcmVwbHkrMDAwMGNmMjNhMGRjMjEzMDJmNTM0NWE3ZjcxNmE4NDlkYTRjMjVmMGQzMDA5NTUwOTJjZjAwMDAwMDAxMTA5ZGJiOTc5MmExNjljZTAxNjA2ZGJiQHJlcGx5LmdpdGh1Yi5jb216_tZAFK7yoQ

Best regards,

Team MEGA
